### DISCUSSION TITLE
AsyncSession automatically switches back to public schema

### DISCUSSION THREAD

**Discussion by tariqjamal-abdulazeez** [sqlalchemy/sqlalchemy#11790] [Usage Questions]
We have a multi-tenancy application where we manipulate data based on tenant schemas. However, after committing, the async session sometimes automatically switches back to the public schema.

**Comment by CaselIT** [comment]
Hi,

Commit will release the connection back to the pool, then a new connection may be obtained from it if more operations are done on the session.

How do you change your schema in your application?

**Reply by tariqjamal-abdulazeez** [reply]
> How do you change your schema in your application?

`await session.execute(text('SET search_path TO tenant_schema'))`

**Reply by CaselIT** [reply]
This seems to imply postgresql.

You could probably look into moving that command to the event https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.after_begin (taking care to use the connection to execute it)

You may use the session.info dict to share what the tenant it with the event

**Reply by tariqjamal-abdulazeez** [reply]
We're implementing this in Fastapi and PostgresSQL DB, so for each request the tenant may vary

**Reply by CaselIT** [reply]
an alternative is to bind the session on a specific connection in fast api. see for example the get session used in this comment https://github.com/sqlalchemy/sqlalchemy/discussions/11789#discussioncomment-10464914 (you can ignore the rest though, since it's not related this discussion otherwise)
